src/command/lsp/mod.rs

```diff
@@ -0,0 +1,230 @@
+        .lazily_resolve_subgraphs(&supergraph_content_root)
+        .await?;
+
+    // Build the service to spin up the language server
+    let (service, socket, _receiver) = ApolloLanguageServer::build_service(
+        Config {
+            root_uri: supergraph_content_root.to_string(),
+            enable_auto_composition: false,
+            force_federation: false,
+            disable_telemetry: false,
+        },
+        HashMap::from_iter(
+            lazily_resolved_supergraph_config
+                .subgraphs()
+                .iter()
+                .map(|(a, b)| (a.to_string(), b.schema().clone())),
+        ),
+    );
+
+    let language_server = service.inner().clone();
+
+    // Spawn a separate thread to handle composition and passing that data to the language server
+    tokio::spawn(async move {
+        // Create a supergraph binary
+        // TODO: Check defaulting behaviour here and see if we need to centralise
+        let federation_version = lazily_resolved_supergraph_config
+            .federation_version()
+            .clone()
+            .unwrap_or(FederationVersion::LatestFedTwo);
+
+        // TODO: Let the supergraph binary exist inside its own task that can respond to being re-installed etc.
+        let supergraph_binary =
+            InstallSupergraph::new(federation_version.clone(), client_config.clone())
+                .install(
+                    None,
+                    lsp_opts.plugin_opts.elv2_license_accepter,
+                    lsp_opts.plugin_opts.skip_update,
+                )
+                .await?;
+
+        // Spin up Runner
+        let mut stream = Runner::default()
+            .setup_subgraph_watchers(
+                lazily_resolved_supergraph_config.subgraphs().clone(),
+                &lsp_opts.plugin_opts.profile,
+                &client_config,
+                500,
+            )
+            .setup_supergraph_config_watcher(lazily_resolved_supergraph_config.clone())
+            .setup_composition_watcher(
+                lazily_resolved_supergraph_config
+                    .fully_resolve_subgraphs(&client_config, &studio_client)
+                    .await
+                    .map_err(ResolveSupergraphConfigError::ResolveSubgraphs)?,
+                supergraph_binary,
+                TokioCommand::default(),
+                FsReadFile::default(),
+                FsWriteFile::default(),
+                OutputTarget::Stdout,
+                Utf8PathBuf::try_from(temp_dir())?,
+            )
+            .run();
+
+        let supergraph_yaml_url =
+            tower_lsp::lsp_types::Url::from_file_path(supergraph_yaml_path)
+                .map_err(|_| anyhow!("Failed to convert supergraph yaml path to url"))?;
+
+        while let Some(event) = stream.next().await {
+            match event {
+                CompositionEvent::Started => {
+                    language_server.composition_did_start().await;
+                }
```

Comment:
there's some interesting stuff the LSP is doing; this sends a `WorkDoneProgressBegin` event that has a corresponding `WorkDoneProgressEnd` event that gets emitted from `composition_did_update`. I feel like that's some important context worth calling out: rover _must_ always send that `WorkDoneProgressBegin` event before the `WorkDoneProgressEnd` event (according to the [LSP docs](https://docs.rs/apollo-language-server/0.3.0/apollo_language_server/struct.ApolloLanguageServer.html)), but I'm not really sure how strong of a 'must' that is

[here's](https://docs.rs/apollo-language-server/0.3.0/src/apollo_language_server/server.rs.html#406) where that method bottoms out, insofar as I can tell. It doesn't look especially dangerous to not emit the right event before emitting the `WorkDoneProgressEnd` event, but there's probably something in the LSP that would get upset about it; either way, might be worth calling out that there's a required-yet-implicit ordering so that future devs don't think we can ignore `CompositionEvent::Started` or something silly like that